### PR TITLE
build: remove confusing and inconsistent disable-asm option

### PR DESCRIPTION
**Pull request by theuni**
1. It didn't actually disable asm usage in our code. Regardless of the setting, asm is used in random.cpp and support/cleanse.cpp.
2. The value wasn't forwarded to libsecp as a user might have reasonably expected.
3. We now have the DISABLE_OPTIMIZED_SHA256 define which is what disable-asm actually did in practice.

If there is any desire, we can hook DISABLE_OPTIMIZED_SHA256 up to a new configure option that actually does what it says.

Additionally, this is one of the last (THE last?) remaining uses of autoconf defines in our crypto code. As such it seems like low-hanging fruit.

### CODE REVIEW

**Comment by theuni** [comment]
Ping @hebasto @fanquake @TheCharlatan for buildsystem eyes.
Ping @sipa for crypto concept ACK/NACK.

**Comment by theuni** [comment]
Also ping @dergoegge for fuzzer interactions ACK/NACK.

**Comment by TheCharlatan** [comment]
Concept ACK

**Comment by dergoegge** [comment]
Concept ACK

We only use `--with-asm=no` for the MSan builds in oss-fuzz but that only applies to secp, so we should be fine on that front.

Our docs (developer-notes.md and fuzzing.md) mention the use of `--disable-asm` to avoid address sanitizer failures. I've not used `--disable-asm` in any of my fuzzing and have not observed any ASan failures but they might depend on the specific sha256 optimizations used? Could also just be a relic of the past...

> If there is any desire, we can hook DISABLE_OPTIMIZED_SHA256 up to a new configure option that actually does what it says.

I think we're good for now but it probably makes sense if anyone runs into the ASan failures mentioned above.

**Comment by theuni** [comment]
@dergoegge Ah, thanks for the correction. The option does indeed currently pass through to secp, so that bullet-point in my commit message is wrong and I'll update it.

It still forwards just fine with our option removed though, so I don't think that changes anything here.

**Comment by dergoegge** [comment]
@jonatack see https://github.com/bitcoin/bitcoin/issues/29178 (we prefer having this broken over reverting the change that introduced it for some reason)

I'm not sure if the `random` harness would trigger the ASan error, the docs mention sha256 which `random` does not use afaict. Could you try `txorphan` or a different harness that does some hashing?

Did you set `UBSAN_OPTIONS` to include `halt_on_error=1`? that is likely why it keeps running if you didn't

**Comment by epiccurious** [comment]
Concept ACK 0eec878e514c23d33d2cd81c44566ea601eb263f.

**Review by sedited — COMMENTED** [review]
Do the mentiones from the `doc/fuzzing.md` to `--disable-asm` still need to be removed? Seems like `--disable-asm` is not recognized by libsecp, it supports a package-specific asm option.

**Comment by theuni** [comment]
> @dergoegge Ah, thanks for the correction. The option does indeed currently pass through to secp, so that bullet-point in my commit message is wrong and I'll update it.

Sorry for the back-and-forth, but I confused myself here. `--disable-asm` doesn't have any effect on libsecp, but `--without-asm` does. No need to update the PR description.

> 
> It still forwards just fine with our option removed though, so I don't think that changes anything here.

I got this wrong but it still accidentally made sense, heh. To be clear, passing `--without-asm` will still do the same as before, which is to disable asm-optimized libsecp without affecting Core at all. It was never "forwarded" in the sense that we interpreted and copied it to libsecp, rather it passed through untouched and still will.

All that confusion just seems like more justification for removal imo :)

**Comment by theuni** [comment]
> The one difference I notice without `--disable-asm` is this additional entry near the beginning of the fuzz run (after which the fuzzer continues to run):
> 
@jonatack Since you can reproduce, any interest in testing this theory https://github.com/bitcoin-core/crc32c-subtree/pull/6#issuecomment-1936095430 ?

**Comment by fanquake** [comment]
Concept ACK

**Review by hebasto — COMMENTED** [review]
Concept ACK.

File: src/Makefile.am [hunk 1]
```diff
@@ -50,10 +50,8 @@ LIBBITCOIN_WALLET_TOOL=libbitcoin_wallet_tool.a
 endif
 
 LIBBITCOIN_CRYPTO = $(LIBBITCOIN_CRYPTO_BASE)
-if USE_ASM
 LIBBITCOIN_CRYPTO_SSE4 = crypto/libbitcoin_crypto_sse4.la
```
Comment (hebasto) [inline src/Makefile.am:53]: It seems there are no reasons to keep `libbitcoin_crypto_sse4`  library anymore. The `crypto/sha256_sse4.cpp` source might be a part of the `libbitcoin_crypto_base`, no?

Comment (theuni) [reply src/Makefile.am:53]: Agreed, but would you be ok with fixing this up in a follow-up? I think it'd make sense to address while dealing with this https://github.com/bitcoin/bitcoin/pull/29404#pullrequestreview-1870581585

**Comment by hebasto** [comment]
> Our docs (developer-notes.md and fuzzing.md) mention the use of `--disable-asm` to avoid address sanitizer failures.

In the libsecp repo, the `--with-asm=no` option is still required to avoid MSan warnings. If this is a common issue for builtin assembly code, it seems possible that the `sha256_sse4.cpp` might cause similar issues in the future without an option to disable it.

Maybe gate the `sha256_sse4::Transform` function with `#if !defined(DISABLE_OPTIMIZED_SHA256)`?

**Comment by theuni** [comment]
@fanquake Do you think it's too late for this for v27? It would imply require getting https://github.com/bitcoin-core/crc32c-subtree/pull/6 in, I think.

**Comment by theuni** [comment]
> In the libsecp repo, the `--with-asm=no` option is still required to avoid MSan warnings.

Can you point to where this is an issue? It looks as though secp has proper `__msan_*` annotations in place these days.

**Comment by fanquake** [comment]
> @fanquake Do you think it's too late for this for v27?

I think it's still possible to do this for v27. It's not clear to me that the other change is strictly a requirement for doing this change?, but I also think that could go in.

**Comment by theuni** [comment]
> > > In the libsecp repo, the `--with-asm=no` option is still required to avoid MSan warnings.
> > 
> > 
> > Can you point to where this is an issue? It looks as though secp has proper `__msan_*` annotations in place these days.
> 
> Sure. For example, [bitcoin-core/secp256k1#1169 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1169#issuecomment-1370003449). It holds for the current libsecp master branch @ [bitcoin-core/secp256k1@0653a25](https://github.com/bitcoin-core/secp256k1/commit/0653a25d50f67c68bd2d196ecc7eddab067d95ef) as well.

Fixed here https://github.com/bitcoin-core/secp256k1/pull/1496 :)

**Comment by theuni** [comment]
Any other sanitizer false-positives while we're at it? :)

**Comment by fanquake** [comment]
> Any other sanitizer false-positives while we're at it? :)

Anything in `test/sanitizer_suppressions/*` is up for grabs (if it can be fixed)

> It would imply require getting https://github.com/bitcoin-core/crc32c-subtree/pull/6 in, I think.

Had another look, and this is correct. We need to merge a fix in the subtree, then do a subtree update here, and then this can be merged, otherwise we'll be left with at least one bug without any workaround. This PR should also be updated to removed the `--disable-asm` mentions in `fuzzing.md` and `developer-notes.md`.

**Comment by fanquake** [comment]
> I think there should be no suppressions related to asm in those files

I'm just talking about general suppressions, nothing asm specific.

**Comment by theuni** [comment]
From `developer-notes.md`:

> There are a number of known problems when using the `address` sanitizer. The
> address sanitizer is known to fail in
> [sha256_sse4::Transform](/src/crypto/sha256_sse4.cpp) which makes it unusable
> unless you also use `--disable-asm` when running configure. We would like to fix
> sanitizer issues, so please send pull requests if you can fix any errors found
> by the address sanitizer (or any other sanitizer).

I'm unable to hit this locally. Presumably if it's still an issue we should be able to fix it with some notations.

Does anyone happen to have a test-case for it?

**Comment by theuni** [comment]
I pushed a commit to remove references to `disable-asm` from our docs as @fanquake suggested, in anticipation of:
- merging https://github.com/bitcoin-core/crc32c-subtree/pull/6 and syncing our subtree
- confirmation that `sha256_sse4::Transform` is no longer an issue and/or supplying a fix for it

**Comment by Empact** [comment]
Concept ACK

**Comment by fanquake** [comment]
> This now depends on #29493

Want to rebase now?

**Comment by theuni** [comment]
Rebased and removed RFC from PR title.

**Review by fanquake — APPROVED** [review]
ACK f8a06f7a02be83e9b76a1b31f1b66a965dbedfce

File: src/Makefile.am (hunk 1 above)
Comment (fanquake) [reply src/Makefile.am:53]: See #29528.

**Comment by achow101** [comment]
I used `--disable-asm` in order to even be able to build when configured for fuzzing. Otherwise, I get this compiler error:

```
crypto/sha256_sse4.cpp:44:9: error: expected relocatable expression
   44 |         "shl    $0x6,%2;"
      |         ^
<inline asm>:1:15894: note: instantiated into assembly here
```

Since crypto/sha256_sse4.cpp is not guarded by DISABLE_OPTIMIZED_SHA256, there's no way for me to compile for fuzzing now.

**Comment by fanquake** [comment]
> Otherwise, I get this compiler error:

Can you open a new issue, with the complete config.log / build output?

**Comment by theuni** [comment]
> Since crypto/sha256_sse4.cpp is not guarded by DISABLE_OPTIMIZED_SHA256, there's no way for me to compile for fuzzing now.

Still trying to understand the problem here, but in the meantime I'll PR an addition of the `DISABLE_OPTIMIZED_SHA256` guard.

**Pull request merged**